sql: add the mapping table for converting oracle functions to tidb functions (#11432)
pingcap/docs-cn: oracle-functions-to-tidb.md

| 获取字符串长度 | `LENGTH(str)` | `CHAR_LENGTH(str)` | |
| 截取字符串 | `SUBSTR('abcdefg',0,2)` => ab<br/> `SUBSTR('abcdefg',1,2)` => ab | `SUBSTRING('abcdefg',0,2)` => 空<br/>`SUBSTRING('abcdefg',1,2)` => ab | Oracle 中起始位置 0 与 1 作用一样，TiDB 中 0 开始截取为空，若需从头开始截全，则应从 1 开始，TiDB 支持 SUBSTRING 和 SUBSTR 函数，作用相同，不用修改。但是注意 TiDB 的下表必须从 1 开始。 |
| 查找字符串 | `INSTR('abcdefg','b',1,1)` | `INSTR('abcdefg','b')` | 从字符串 'abcdefg' 第一个字符开始查询，返回 'b' 字符串第一次出现的位置。 |
| 查找字符串 | `INSTR('stst','s',1,2)` | `LENGTH(SUBSTRING_INDEX('stst','s',2)) + 1` | 从字符串 'stst' 第一个字符开始查找，返回 's' 字符第二次出现的位置。 |
| 查找字符串 | `INSTR('abcabc','b',2,1)` | `LOCATE('b','abcabc',2)` | 从字符串 `abcabc` 第二个字符开始查询，返回 `b` 字符第一次出现的位置。 |
| 获取日期之间间隔月份 | `MONTHS_BETWEEN(ENDDATE,SYSDATE)` | `TIMESTAMPDIFF(MONTH,SYSDATE,ENDDATE)` | Oracle 中 `MONTHS_BETWEEN()` 函数与 TiDB 中 `TIMESTAMPDIFF()` 函数的结果会有误差。`TIMESTAMPDIFF()` 只保留整数月。使用时需要注意，两个函数的参数位置相反。 |
| 列合并为行 | `LISTAGG(CONCAT(E.dimensionid,'---',E.DIMENSIONNAME),'***') within GROUP(ORDER BY  DIMENSIONNAME)` | `GROUP_CONCAT(CONCAT(E.dimensionid,'---',E.DIMENSIONNAME) ORDER BY DIMENSIONNAME SEPARATOR '***')` | 将一列字段合并为一行并根据 `***` 符号进行分割。 |
| 获取当前时间（精确到微秒）| `SYSTIMESTAMP` | `CURRENT_TIMESTAMP(6)` | |

> | ASCII 值转化为对应字符 | `CHR(n)` | `CHAR(n)` | Oracle 中制表符 (`CHR(9)`)、换行符 
>  (`CHR(10)`)、回车符 (`CHR(13)`) 对应 TiDB 中的 `CHAR(9)`、`CHAR(10)`、`CHAR(13)`。 |

Author: ```suggestion
| ASCII 值转化为对应字符 | `CHR(n)` | `CHAR(n)` | Oracle 中制表符 (`CHR(9)`)、换行符 (`CHR(10)`)、回车符 (`CHR(13)`) 对应 TiDB 中的 `CHAR(9)`、`CHAR(10)`、`CHAR(13)`。 |
```

| 获取字符串长度 | `LENGTH(str)` | `CHAR_LENGTH(str)` | |
| 截取字符串 | `SUBSTR('abcdefg',0,2)` => ab<br/> `SUBSTR('abcdefg',1,2)` => ab | `SUBSTRING('abcdefg',0,2)` => 空<br/>`SUBSTRING('abcdefg',1,2)` => ab | Oracle 中起始位置 0 与 1 作用一样，TiDB 中 0 开始截取为空，若需从头开始截全，则应从 1 开始，TiDB 支持 SUBSTRING 和 SUBSTR 函数，作用相同，不用修改。但是注意 TiDB 的下表必须从 1 开始。 |
| 字符查找函数 INSTR | `INSTR('abcdefg','b',1,1)` | `INSTR('abcdefg','b')` | 从字符串 'abcdefg' 第 1 个字符开始查询，返回 'b' 字符串第 1 次出现按的位置。 |
| 字符查找函数 INSTR | `INSTR('stst','s',1,2)` | `LENGTH(SUBSTRING_INDEX('stst','s',2))+1` | 从字符串 'stst' 第一个字符开始查找，返回 's' 字符第 2 次出现的位置，查找非第一次出现的位置时使用。 |
| 字符查找函数 | `INSTR('abcabc','b',2,1)` | `LOCATE('b','abcabc',2)` | 字符查找函数。从字符串 `abcabc` 第 2 个字符开始查询，返回 `b` 字符串第 1 次出现按的位置。 |
| 获取日期之间间隔月份 | `MONTHS_BETWEEN(enddate,sysdate)` | `TIMESTAMPDIFF(MONTH,sysdate,enddate)` |  `MONTHS_BETWEEN` 函数返回两个日期之间的月份值，可以用 `TIMESTAMPDIFF` 替换，但是结果上会有误差，`TIMESTAMPDIFF` 只保留整数月，应该测试后按照业务看是否替换，另外注意参数位置相反。 |

> | 列转行函数 | `LISTAGG(concat(E.dimensionid,'---',E.DIMENSIONNAME),'***') within GROUP(ORDER BY  DIMENSIONNAME)` | `GROUP_CONCAT(concat(E.dimensionid,'---',E.DIMENSIONNAME) ORDER BY DIMENSIONNAME SEPARATOR '***')` | Oracle 中的 `LISTAGG` 需要改写为 TiDB 的 `GROUP_CONCAT` 函数；将一列字段合并为一行并根据 `***` 符号进行分割。 |

Reviewer: ```suggestion
| 列合并为行 | `LISTAGG(CONCAT(E.dimensionid,'---',E.DIMENSIONNAME),'***') within GROUP(ORDER BY  DIMENSIONNAME)` | `GROUP_CONCAT(CONCAT(E.dimensionid,'---',E.DIMENSIONNAME) ORDER BY DIMENSIONNAME SEPARATOR '***')` | 将一列字段合并为一行并根据 `***` 符号进行分割。 |
```

| 获取当前时间（精确到微秒）| `SYSTIMESTAMP` | `CURRENT_TIMESTAMP(6)` | |
| ASCII 值转化为对应字符 | `CHR(n)` | `CHAR(n)` | Oracle 中制表符 `CHR(9)`/换行符 `CHR(10)`/回车符 `CHR(13)` 对应 TiDB 中的 `CHAR(9)`/`CHAR(10)`/`CHAR(13)`。 |

## 语法差异对照

本节介绍 Oracle 部分语法与 TiDB 的差异。

> ### 删除语句中对表加别名

Reviewer: ```suggestion
### 删除语句中为表加别名
```

> Oracle：`DELETE FROM test t WHERE t.xxx = xxx;`
> 
> TiDB： `DELETE FROM test WHERE xxx = xxx;`
> 
> 说明：TiDB 不支持删除语句中为表起别名。

Reviewer: ```suggestion
TiDB 不支持在删除语句中对表加别名，即 Oracle 中 `DELETE FROM test t WHERE t.xxx = xxx;` 删除语句应改写为 `DELETE FROM test WHERE xxx = xxx;`。

```

> ### 字符串引号识别区别

Reviewer: ```suggestion
### 字符串语法
```

> Oracle：'a'
> 
> TiDB：'a' / "a"
> 
> 说明：Oracle 只能识别单引号，TiDB 能识别单引号与双引号。

Reviewer: ```suggestion
- Oracle 中字符串只能使用单引号 ('')，例如 `'a'`。
- TiDB 中字符串可以使用单引号 ('') 或双引号 ("")，例如 `'a'` 或 `"a"`。
```

### NULL 与空串区别

> Oracle：不区分 NULL 和 ''
> 
> TiDB：区分 NULL 和 ''
> 
> 说明：Oracle 空串就是 NULL，TiDB 需要把空串转换为 NULL 数据。TiDB 中 NULL 和 '' 是有区别的。

Reviewer: ```suggestion
- Oracle 不区分 `NULL` 和空字符串 `''`，即 `NULL` 与 `''` 是等价的。
- TiDB 区分 `NULL` 和空字符串 `''`，TiDB 需要将空字符串转换为 `NULL`。
```

> ### 插入本表中查询出来的数据

Reviewer: ```suggestion
### `INSERT` 语句中读写同一张表
```

> Oracle：`INSERT INTO table1 VALUES (feild1,(SELECT feild2 FROM table1 WHERE...))`
> 
> TiDB：`INSERT INTO table1 VALUES（feild1,(SELECT T.fields2 FROM table1 T WHERE...)`
> 
> 说明：TiDB 不支持在同一个表中先查这个表再更新该表。

Reviewer: ```suggestion
TiDB 不支持在同一个表中先查询再更新该表。即 Oracle 中 `INSERT INTO table1 VALUES (feild1,(SELECT feild2 FROM table1 WHERE...))` 语句在 TiDB 中应改写为 `INSERT INTO table1 VALUES（feild1,(SELECT T.fields2 FROM table1 T WHERE...)`。
```

> 说明：TiDB 不支持在同一个表中先查这个表再更新该表。

Author: TiDB 给的这个例子是支持呢还是不支持的呢？

> ### 限制取前 n 条数据

Reviewer: ```suggestion
### 获取前 n 条数据
```

> Oracle：`ROWNUM <= 10`
> 
> TiDB：`LIMIT 10`
> 
> 说明：Oracle 中的 `ROWNUM` 可以使用 `LIMIT` 等价代替。Hibernate Query Language (HQL) 方式运行带 `LIMIT` 的 SQL 语句会出现错误，需要将 Hibernate 的运行方式改为 SQL 方式运行。

Reviewer: ```suggestion
- Oracle：通过 `ROWNUM <= n` 获取前 n 条数据。
- TiDB：通过 `LIMIT n` 获取前 n 条数据。Hibernate Query Language (HQL) 方式运行带 `LIMIT` 的 SQL 语句会出现错误，需要将 Hibernate 的运行方式改为 SQL 方式运行。
```

### UPDATE 语句

> Oracle：`UPDATE test1 SET(test1.name,test1.age)= (SELECT test2.name,test2.age FROM test2 WHERE test2.id=test1.id)`
> 
> TiDB：`UPDATE test1,test2 SET test1.name=test2.name,test1.age=test2.age WHERE test1.id=test2.id`
> 
> 说明：TiDB 在多表更新的时候，需要在 `SET` 的时候把具体的字段更新关系都列出来。

Reviewer: ```suggestion
TiDB 在多表更新的时候，需要在 `SET` 时列出具体的字段更新关系。即 Oracle 中 `UPDATE test1 SET(test1.name,test1.age)= (SELECT test2.name,test2.age FROM test2 WHERE test2.id=test1.id)` 更新语句在 TiDB 中应改写为 `UPDATE test1,test2 SET test1.name=test2.name,test1.age=test2.age WHERE test1.id=test2.id`。
```

### 派生表别名

Oracle：`SELECT * FROM (SELECT * FROM test)`

TiDB：`SELECT * FROM (SELECT * FROM test)` t

说明：TiDB 多表查询的时候，每一个派生出来的表都必须有一个自己的别名。

### 差集运算

Oracle：`SELECT * FROM t1 MINUS SELECT * FROM t2;`

TiDB：`SELECT * FROM t1 EXCEPT SELECT * FROM t2;`

说明：TiDB 不支持 `MINUS`，需要改写为 `EXCEPT`。

### 空值取别名

Oracle： `SELECT null AS xx FROM dual`

TiDB：`SELECT '' AS xx FROM dual`

说明：NULL 与 '' 在 TiDB 中含义不同。Oracle 中 NULL 在 TiDB 中需要改写为 ''。

> ### 注释

Reviewer: ```suggestion
### 注释语法
```

Oracle：`--注释`

TiDB： `-- 注释`

> 说明：Oracle 的 `--` 后面不需要空格，TiDB 的 `--` 后面需要有一个空格。

Author: 那个说明还是保留吧？要不然不容易看出区别。

---
title: Oracle 与 TiDB 函数和语法差异对照
summary: 了解 Oracle 与 TiDB 函数和语法差异对照。
---

# Oracle 与 TiDB 函数和语法差异对照

本文档提供了 Oracle 与 TiDB 的函数和语法差异对照，方便你根据 Oracle 函数查找对应的 TiDB 函数，了解 Oracle 与 TiDB 语法差异。

> > **说明：**

Reviewer: ```suggestion
> **注意：**
```

| 获取日期之间间隔月份 | `MONTHS_BETWEEN(ENDDATE,SYSDATE)` | `TIMESTAMPDIFF(MONTH,SYSDATE,ENDDATE)` | Oracle 中 `MONTHS_BETWEEN()` 函数与 TiDB 中 `TIMESTAMPDIFF()` 函数的结果会有误差。`TIMESTAMPDIFF()` 只保留整数月。使用时需要注意，两个函数的参数位置相反。 |
| 列合并为行 | `LISTAGG(CONCAT(E.dimensionid,'---',E.DIMENSIONNAME),'***') within GROUP(ORDER BY  DIMENSIONNAME)` | `GROUP_CONCAT(CONCAT(E.dimensionid,'---',E.DIMENSIONNAME) ORDER BY DIMENSIONNAME SEPARATOR '***')` | 将一列字段合并为一行并根据 `***` 符号进行分割。 |
| 获取当前时间（精确到微秒）| `SYSTIMESTAMP` | `CURRENT_TIMESTAMP(6)` | |
| ASCII 值转化为对应字符 | `CHR(n)` | `CHAR(n)` | Oracle 中制表符 (`CHR(9)`)、换行符 (`CHR(10)`)、回车符 (`CHR(13)`) 对应 TiDB 中的 `CHAR(9)`、`CHAR(10)`、`CHAR(13)`。 |

## 语法差异

本节介绍 Oracle 部分语法与 TiDB 的差异。

### 字符串语法

- Oracle：字符串只能使用单引号 ('')。例如 `'a'`。
- TiDB：字符串可以使用单引号 ('') 或双引号 ("")。例如 `'a'` 或 `"a"`。

### `NULL` 与空字符串的区分

- Oracle：不区分 `NULL` 和空字符串 `''`，即 `NULL` 与 `''` 是等价的。
- TiDB：区分 `NULL` 和空字符串 `''`。

### `INSERT` 语句中读写同一张表

Oracle 支持 `INSERT` 语句中读写同一张表。例如:

```sql
INSERT INTO table1 VALUES (feild1,(SELECT feild2 FROM table1 WHERE...))
```

TiDB 不支持 `INSERT` 语句中读写同一张表。例如：

```sql
INSERT INTO table1 VALUES (feild1,(SELECT T.fields2 FROM table1 T WHERE...))
```

> ### 获取前 n 条数据

Author: ```suggestion
### 获取前 n 行数据
```

> - Oracle：通过 `ROWNUM <= n` 获取前 n 条数据。例如，`ROWNUM <= 10`。
> - TiDB：通过 `LIMIT n` 获取前 n 条数据。例如，`LIMIT 10`。Hibernate Query Language (HQL) 方式运行带 `LIMIT` 的 SQL 语句会出现错误，需要将 Hibernate 的运行方式改为 SQL 方式运行。

Author: ```suggestion
- Oracle：通过 `ROWNUM <= n` 获取前 n 行数据。例如，`ROWNUM <= 10`。
- TiDB：通过 `LIMIT n` 获取前 n 行数据。例如，`LIMIT 10`。Hibernate Query Language (HQL) 方式运行带 `LIMIT` 的 SQL 语句会出现错误，需要将 Hibernate 的运行方式改为 SQL 方式运行。
```

| 截取数据 | `TRUNC(2.136) = 2`<br/> `TRUNC(2.136,2) = 2.13` | `TRUNCATE(2.136,0) = 2`<br/> `TRUNCATE(2.136,2) = 2.13` | 数据精度保留，直接截取相应小数位，不涉及四舍五入。 |
| 获取序列下一个值 | `sequence_name.NEXTVAL` | `NEXTVAL(sequence_name)` | |
| 获取随机序列值 | `SYS_GUID()` | `UUID()` | TiDB 返回一个通用唯一识别码 (UUID)。|
| 左/右外连接 | `SELECT * FROM a, b WHERE a.id = b.id(+);`<br/>`SELECT * FROM a, b WHERE a.id(+) = b.id;` | `SELECT * FROM a LEFT JOIN b ON a.id = b.id;`<br/>`SELECT * FROM a RIGHT JOIN b ON a.id = b.id;` | 关联查询时，TiDB 不支持使用 (+) 实现左/右关联，只能通过 `LEFT JOIN` 或 `RIGHT JOIN` 实现。|
| `NVL()` | `NVL(key,val)` | `IFNULL(key,val)` | 如果该字段值为 `NULL`，则返回 val 值，否则返回该字段的值。 |
| `NVL2()` | `NVL2(key, val1, val2)`  | `IF(key is NULL, val1, val2)` | 如果该字段值非 `NULL`，则返回 val1 值，否则返回 val2 值。|
| `DECODE()` | <li>`DECODE(key,val1,val2,val3)`</li><li>`DECODE(value,if1,val1,if2,val2,...,ifn,valn,val)`</li> | <li>`IF(key=val1,val2,val3)`</li><li>`CASE WHEN value=if1 THEN val1 WHEN value=if2 THEN val2,...,WHEN value=ifn THEN valn ELSE val END`</li> | <li>如果该字段值等于 val1，则返回 val2，否则返回 val3。</li><li>当该字段值满足条件 1 (if1) 时，返回 val1，满足条件 2 (if2) 时，返回 val2，满足条件 3 (if3) 时，返回 val3。</li> |

> | 拼接字符串 `a` 和 `b` | `'a' \|\| 'b'` | `CONCAT('a','b')` | |

Reviewer: ```suggestion
| 拼接字符串 `a` 和 `b` | `'a' || 'b'` | `CONCAT('a','b')` | |
```

> | 拼接字符串 `a` 和 `b` | `'a' \|\| 'b'` | `CONCAT('a','b')` | |

Reviewer: ```suggestion
| 拼接字符串 `a` 和 `b` | <code>'a' \|\| 'b'</code>  | `CONCAT('a','b')` | |
```

---
title: Oracle 与 TiDB 函数和语法差异对照
summary: 了解 Oracle 与 TiDB 函数和语法差异对照。
---

# Oracle 与 TiDB 函数和语法差异对照

本文档提供了 Oracle 与 TiDB 的函数和语法差异对照，方便你根据 Oracle 函数查找对应的 TiDB 函数，了解 Oracle 与 TiDB 语法差异。

> **说明：**
>
> 本文的内容是基于 Oracle 12.2.0.1.0 和 TiDB v5.4.0，其他版本可能存在差异。

## 函数对照表

下表列出了 Oracle 与 TiDB 部分函数的对照表。

| 函数 | Oracle 语法 | TiDB 语法 | 说明 |
|---|---|---|---|

> | 转换数据类型 | <li>`TO_NUMBER(key)`</li><li>`TO_CHAR(key)`</li> | `CONVERT(key,dataType)` | TiDB 支持转换 `BINARY`、`CHAR`、`DATE`、`DATETIME`、`TIME`、`SIGNED INTEGER`、`UNSIGNED INTEGER` 和 `DECIMAL` 类型。 |

Author: ```suggestion
| 转换数据类型 | <li>`TO_NUMBER(key)`</li><li>`TO_CHAR(key)`</li> | `CONVERT(key,dataType)` | TiDB 支持转换为下面类型：`BINARY`、`CHAR`、`DATE`、`DATETIME`、`TIME`、`SIGNED INTEGER`、`UNSIGNED INTEGER` 和 `DECIMAL`。 |
```

| 日期类型转换为字符串类型 | <li>`TO_CHAR(SYSDATE,'yyyy-MM-dd hh24:mi:ss')`</li> <li>`TO_CHAR(SYSDATE,'yyyy-MM-dd')`</li> | <li>`DATE_FORMAT(NOW(),'%Y-%m-%d %H:%i:%s')`</li><li>`DATE_FORMAT(NOW(),'%Y-%m-%d')`</li> | TiDB 的格式化字符串大小写敏感。 |
| 字符串类型转换为日期类型 | <li>`TO_DATE('2021-05-28 17:31:37','yyyy-MM-dd hh24:mi:ss')`</li><li>`TO_DATE('2021-05-28','yyyy-MM-dd hh24:mi:ss')`</li> | <li>`STR_TO_DATE('2021-05-28 17:31:37','%Y-%m-%d %H:%i:%s')`</li><li>`STR_TO_DATE('2021-05-28','%Y-%m-%d%T')` </li> | TiDB 的格式化字符串大小写敏感。 |
| 获取系统当前时间（精确到秒）| `SYSDATE` | `NOW()` | |
| 获取两个日期 `date1` 和 `date2` 相差的天数 | `date1 - date2` | `DATEDIFF(date1, date2)` | |
| 日期增加/减少 n 天 | `DATEVAL + n` | `DATE_ADD(dateVal,INTERVAL n DAY)` | `n` 可为负数。|
| 日期增加/减少 n 月 | `ADD_MONTHS(dateVal,n)`| `DATE_ADD(dateVal,INTERVAL n MONTH)` | `n` 可为负数。|
| 获取日期到日 | `TRUNC(SYSDATE)` | <li>`CAST(NOW() AS DATE)`</li><li>`DATE_FORMAT(NOW(),'%Y-%m-%d')`</li> | TiDB 中 `CAST` 与 `DATE_FORMAT` 结果一致。|

> | 获取日期的月份 | `TRUNC(SYSDATE,'mm')` | `DATE_ADD(CURDATE(),interval - day(CURDATE()) + 1 day)`  | |

Author: ```suggestion
| 获取日期当月第一天 | `TRUNC(SYSDATE,'mm')` | `DATE_ADD(CURDATE(),interval - day(CURDATE()) + 1 day)`  | |
```

> | 向下取整 | `TRUNC(2.136) = 2`<br/> `TRUNC(2.136,2) = 2.14` | `TRUNCATE(2.136,0) = 2`<br/> `TRUNCATE(2.136,2) = 2.14` |  |

Author: ```suggestion
| 截取数据 | `TRUNC(2.136) = 2`<br/> `TRUNC(2.136,2) = 2.14` | `TRUNCATE(2.136,0) = 2`<br/> `TRUNCATE(2.136,2) = 2.14` |  |
```

> | 截取数据 | `TRUNC(2.136) = 2`<br/> `TRUNC(2.136,2) = 2.14` | `TRUNCATE(2.136,0) = 2`<br/> `TRUNCATE(2.136,2) = 2.14` |  |

Author: ```suggestion
| 截取数据 | `TRUNC(2.136) = 2`<br/> `TRUNC(2.136,2) = 2.13` | `TRUNCATE(2.136,0) = 2`<br/> `TRUNCATE(2.136,2) = 2.13` | 数据精度保留，直接截取相应小数位，不涉及四舍五入。 |
```

> | 拼接字符串 `a` 和 `b` | `'a' || 'b'` | `CONCAT('a','b')` | |

Author: ```suggestion
| 拼接字符串 `a` 和 `b` | `'a' \|\| 'b'` | `CONCAT('a','b')` | |
```

| 获取序列下一个值 | `SEQUENCENAME.NEXTVAL` | `NEXTVAL(sequenceName)` | |
| 左/右外连接 | `SELECT * FROM a, b WHERE a.id = b.id(+);`<br/>`SELECT * FROM a, b WHERE a.id(+) = b.id;` | `SELECT * FROM a LEFT JOIN b ON a.id = b.id;`<br/>`SELECT * FROM a RIGHT JOIN b ON a.id = b.id;` | 关联查询时，TiDB 不支持使用 (+) 实现左/右关联，只能通过 `LEFT JOIN` 或 `RIGHT JOIN` 实现。|
| 获取随机序列值 | `SYS_GUID()` | `UUID()` | TiDB 返回一个通用唯一识别码 (UUID)。|
| `NVL()` | `NVL(key,val)` | `IFNULL(key,val)` | 如果该字段值为 `NULL`，则返回 val 值，否则返回该字段的值。 |
| `NVL2()` | `NVL2(key, val1, val2)`  | `IF(key is NULL, val1, val2)` | 如果该字段值非 `NULL`，则返回 val1 值，否则返回 val2 值。|
| `DECODE()` | <li>`DECODE(key,val1,val2,val3)`</li><li>`DECODE(value,if1,val1,if2,val2,...,ifn,valn,val)`</li> | <li>`IF(key=val1,val2,val3)`</li><li>`CASE WHEN value=if1 THEN val1 WHEN value=if2 THEN val2,...,WHEN value=ifn THEN valn ELSE val END`</li> | <li>如果该字段值等于 val1，则返回 val2，否则返回 val3。</li><li>当该字段值满足条件 1 (if1) 时，返回 val1，满足条件 2 (if2) 时，返回 val2，满足条件 3 (if3) 时，返回 val3。</li> |
| 获取字符串长度 | `LENGTH(str)` | `CHAR_LENGTH(str)` | |

> | 获取子串 | `SUBSTR('abcdefg',0,2) = 'ab'`<br/> `SUBSTR('abcdefg',1,2) = 'ab'` | `SUBSTRING('abcdefg',0,2) = ''`<br/>`SUBSTRING('abcdefg',1,2) = 'ab'` | <li>Oracle 中起始位置 0 与 1 作用一样。</li><li>TiDB 中 0 开始获取的子串为空，若需从字符串的起始位置开始，则应从 1 开始。<li> |

Author: ```suggestion
| 获取子串 | `SUBSTR('abcdefg',0,2) = 'ab'`<br/> `SUBSTR('abcdefg',1,2) = 'ab'` | `SUBSTRING('abcdefg',0,2) = ''`<br/>`SUBSTRING('abcdefg',1,2) = 'ab'` | <li>Oracle 中起始位置 0 与 1 作用一样。</li><li>TiDB 中 0 开始获取的子串为空，若需从字符串的起始位置开始，则应从 1 开始。</li> |
```

| 字符串在源字符串中的位置 | `INSTR('abcdefg','b',1,1)` | `INSTR('abcdefg','b')` | 从字符串 'abcdefg' 第一个字符开始查询，返回 'b' 字符串第一次出现的位置。 |
| 字符串在源字符串中的位置 | `INSTR('stst','s',1,2)` | `LENGTH(SUBSTRING_INDEX('stst','s',2)) + 1` | 从字符串 'stst' 第一个字符开始查找，返回 's' 字符第二次出现的位置。 |
| 字符串在源字符串中的位置 | `INSTR('abcabc','b',2,1)` | `LOCATE('b','abcabc',2)` | 从字符串 `abcabc` 第二个字符开始查询，返回 `b` 字符第一次出现的位置。 |
| 获取日期之间间隔月份 | `MONTHS_BETWEEN(ENDDATE,SYSDATE)` | `TIMESTAMPDIFF(MONTH,SYSDATE,ENDDATE)` | Oracle 中 `MONTHS_BETWEEN()` 函数与 TiDB 中 `TIMESTAMPDIFF()` 函数的结果会有误差。`TIMESTAMPDIFF()` 只保留整数月。使用时需要注意，两个函数的参数位置相反。 |
| 列合并为行 | `LISTAGG(CONCAT(E.dimensionid,'---',E.DIMENSIONNAME),'***') within GROUP(ORDER BY  DIMENSIONNAME)` | `GROUP_CONCAT(CONCAT(E.dimensionid,'---',E.DIMENSIONNAME) ORDER BY DIMENSIONNAME SEPARATOR '***')` | 将一列字段合并为一行并根据 `***` 符号进行分割。 |
| 获取当前时间（精确到微秒）| `SYSTIMESTAMP` | `CURRENT_TIMESTAMP(6)` | |
| ASCII 值转化为对应字符 | `CHR(n)` | `CHAR(n)` | Oracle 中制表符 (`CHR(9)`)、换行符 (`CHR(10)`)、回车符 (`CHR(13)`) 对应 TiDB 中的 `CHAR(9)`、`CHAR(10)`、`CHAR(13)`。 |

## 语法差异

本节介绍 Oracle 部分语法与 TiDB 的差异。

> ### 删除语句中为表加别名
> 
> Oracle 支持在删除语句中为表起别名。例如：
> 
> ```sql
> DELETE FROM test t WHERE t.xxx = xxx
> ```
> 
> TiDB 不支持在删除语句中为表起别名。例如
> 
> ```sql
> DELETE FROM test WHERE xxx = xxx
> ```
> 

Author: ```suggestion
```

### 字符串语法

- Oracle：字符串只能使用单引号 ('')。例如 `'a'`。
- TiDB：字符串可以使用单引号 ('') 或双引号 ("")。例如 `'a'` 或 `"a"`。

### `NULL` 与空字符串的区分

- Oracle：不区分 `NULL` 和空字符串 `''`，即 `NULL` 与 `''` 是等价的。

> - TiDB：区分 `NULL` 和空字符串 `''`，TiDB 需要将 `''` 转换为 `NULL`。

Author: ```suggestion
- TiDB：区分 `NULL` 和空字符串 `''`。
```

### `INSERT` 语句中读写同一张表

Oracle 支持 `INSERT` 语句中读写同一张表。例如:

```sql
INSERT INTO table1 VALUES (feild1,(SELECT feild2 FROM table1 WHERE...))
```

TiDB 不支持 `INSERT` 语句中读写同一张表。例如：

```sql
INSERT INTO table1 VALUES (feild1,(SELECT T.fields2 FROM table1 T WHERE...))
```

### 获取前 n 条数据

- Oracle：通过 `ROWNUM <= n` 获取前 n 条数据。例如，`ROWNUM <= 10`。
- TiDB：通过 `LIMIT n` 获取前 n 条数据。例如，`LIMIT 10`。Hibernate Query Language (HQL) 方式运行带 `LIMIT` 的 SQL 语句会出现错误，需要将 Hibernate 的运行方式改为 SQL 方式运行。

### `UPDATE` 语句多表更新

Oracle 多表更新时不需要列出具体的字段更新关系。例如：

```sql
UPDATE test1 SET(test1.name,test1.age) = (SELECT test2.name,test2.age FROM test2 WHERE test2.id=test1.id)
```

TiDB 多表更新时需要在 `SET` 时把具体的字段更新关系都列出来。例如：

```sql
UPDATE test1,test2 SET test1.name=test2.name,test1.age=test2.age WHERE test1.id=test2.id
```

### 派生表别名

Oracle 多表查询时，派生表可以不起别名。例如：

```sql
SELECT * FROM (SELECT * FROM test)
```

TiDB 多表查询时，每一个派生出来的表都必须有一个自己的别名。例如：

```sql
SELECT * FROM (SELECT * FROM test) t
```

### 差集运算

Oracle 使用 `MINUS` 进行差集运算。例如：

```sql
SELECT * FROM t1 MINUS SELECT * FROM t2
```

TiDB 不支持 `MINUS`，需要改写为 `EXCEPT`。例如：

```sql
SELECT * FROM t1 EXCEPT SELECT * FROM t2
```

> ### 空值取别名
> 
> Oracle 中`NULL` 与 `''` 是等价的，为空值起别名时，不需要特殊转换。例如：
> 
> ```sql
> SELECT NULL AS ... FROM DUAL
> ```
> 
> TiDB 中 `NULL` 与 `''` 含义不同，为空值起别名时，需要将 `NULL` 改写为 `''`。例如：
> 
> ```sql
> SELECT '' AS ... FROM DUAL
> ```
> 

Author: ```suggestion
```

---

> title: Oracle 与 TiDB 函数对照表

Reviewer 2: 按照下面的内容，可能更应该叫 ' 函数与语法对照表‘

Author: 好建议，我拆分下

| 获取当前时间，精确到 s | `SYSDATE` | `NOW()` | 获取系统当前时间。 |
| 获取两个日期相差的天数 | `DATE1 - DATE2` | `DATEDIFF(date1, date2)` | 获取 `DATE1 - DATE2` 两个日期之间相差的天数，只能精确到天。 |
| 日期数据 +/- n 天 | `DATEVAL + n` | `DATE_ADD(dateVal,INTERVAL n DAY)` | 日期数据增加 `n` 天，`n` 可为负数 |
| 日期数据 +/- n 月 | `ADD_MONTHS(dateVal,n)`| `DATE_ADD(dateVal,INTERVAL n MONTH)` | 日期数据增加 `n` 月，`n` 可为负数。 |
| `TRUNC()` 函数截取日期到日 | `TRUNC(sysdate)` | <li>`CAST(now() as date)`</li><li>`DATE_FORMAT(now(),'%Y-%m-%d')`</li> | <li>获取时间格式（2019-07-26 00:00:00）的返回值。Oracle 中的 `TRUNC(sysdate)` 只是截取到日，不会截取到时分秒。而 TiDB 中与之对应的截取日写法是 `CAST(now() as date)`。</li><li>CAST 与 `DATE_FORMAT` 结果一致。</li> |
| `TRUNC()` 函数获取日期当月第一天 | `TRUNC(sysdate,'mm')` | `DATE_ADD(curdate(),interval -day(curdate())+1 day)`  | 获取当月第一天。 |
| `TRUNC()` 函数截取数据 | `TRUNC(2.136)` => 2<br/> `TRUNC(2.136,2)` => 2.14 | `TRUNCATE(2.136,0)` => 2<br/> `TRUNCATE(2.136,2)` => 2.14 | 数据精度保留，直接截取相应小数位，不涉及四舍五入。 |
| 字符串连接 | a' \|\| 'b' | `CONCAT('a','b')` | 字符串拼接。 |

> | 删除语句中对表加别名 | `DELETE FROM test t WHERE t.xxx = xxx;` | `DELETE FROM test WHERE xxx = xxx;` | 删除语句，TiDB 不支持删除语句中为表起别名。 |

Reviewer: ```suggestion
| 删除语句中对表加别名 | `DELETE FROM test t WHERE t.xxx = xxx;` | `DELETE FROM test WHERE xxx = xxx;` | TiDB 不支持删除语句中为表起别名。 |
```

| 获取序列下一个值 | `SEQUENCENAME.NEXTVAL` | `NEXTVAL(sequenceName)` | 获取序列的下一个值。 |

> | 左/右外连接 | `SELECT * FROM a, b WHERE a.id = b.id(+);`<br/>`SELECT * FROM a, b WHERE a.id(+) = b.id;` | `SELECT * FROM a LEFT JOIN b ON a.id = b.id;`<br/>`SELECT * FROM a RIGHT JOIN b ON a.id = b.id;`      | 关联查询时，TiDB 不支持使用 (+) 实现左/右关联，只能通过 `LEFT/RIGHT JOIN` 实现。 |

Reviewer: ```suggestion
| 左/右外连接 | `SELECT * FROM a, b WHERE a.id = b.id(+);`<br/>`SELECT * FROM a, b WHERE a.id(+) = b.id;` | `SELECT * FROM a LEFT JOIN b ON a.id = b.id;`<br/>`SELECT * FROM a RIGHT JOIN b ON a.id = b.id;` | 关联查询时，TiDB 不支持使用 (+) 实现左/右关联，只能通过 `LEFT/RIGHT JOIN` 实现。|
```

> | 获取随机序列值 | `SYS_GUID()` | `UUID()` | 返回一个通用唯一识别码 (UUID)。 |

Reviewer: ```suggestion
| 获取随机序列值 | `SYS_GUID()` | `UUID()` | TiDB 返回一个通用唯一识别码 (UUID)。|
```

> | `NVL()` 函数 | `NVL(key,val)` | `IFNULL(key,val)` | 如果该字段值为空，则返回 val 值，否则返回该字段的值。 |

Reviewer: ```suggestion
| `NVL()` 函数 | `NVL(key,val)` | `IFNULL(key,val)` | 如果该字段值为空，则返回 val 值，否则返回该字段的值。 |
```

| `NVL2()` 函数 | `NVL2(key, val1, val2)`  | `if(key is null, val1, val2)` | 如果该字段值非 NULL，则返回 val1 值，否则返回 val2 值。       |
| `DECODE()` 函数 | <li>`DECODE(key,val1,val2,val3)`</li><li>`DECODE(value,if1,val1,if2,val2,...,ifn,valn,val)`</li> | <li>`IF(key=val1,val2,val3)`</li><li>`CASE WHEN value=if1 THEN val1 WHEN value=if2 THEN val2,,,WHEN value=ifn THEN valn ELSE val END`</li> | <li>如果该字段值对于 val1，则返回 val2，否则返回 val3。</li><li>当该字段值等于条件 1 时，返回 val1，等于条件 2 时，返回 val2… </li> |

> | 获取字符串长度 | `LENGTH(str)` | `CHAR_LENGTH(str)` | 获取字符串长度。 |

Reviewer: ```suggestion
| 获取字符串长度 | `LENGTH(str)` | `CHAR_LENGTH(str)` | |
```

> | 截取字符串函数 | `SUBSTR('abcdefg',0,2)` => ab<br/> `SUBSTR('abcdefg',1,2)` => ab | `SUBSTRING('abcdefg',0,2)` => 空<br/>`SUBSTRING('abcdefg',1,2)` => ab | 截取字符串，Oracle 中起始位置 0 与 1 作用一样，TiDB 中 0 开始截取为空，若需从头开始截全，则应从 1 开始，TiDB 支持 SUBSTRING 和 SUBSTR 函数，作用相同，不用修改。但是注意 TiDB 的下表必须从 1 开始。 |

Reviewer: ```suggestion
| 截取字符串 | `SUBSTR('abcdefg',0,2)` => ab<br/> `SUBSTR('abcdefg',1,2)` => ab | `SUBSTRING('abcdefg',0,2)` => 空<br/>`SUBSTRING('abcdefg',1,2)` => ab | Oracle 中起始位置 0 与 1 作用一样，TiDB 中 0 开始截取为空，若需从头开始截全，则应从 1 开始，TiDB 支持 SUBSTRING 和 SUBSTR 函数，作用相同，不用修改。但是注意 TiDB 的下表必须从 1 开始。 |
```

> | 字符查找函数 INSTR | `INSTR('abcdefg','b',1,1)` | `INSTR('abcdefg','b')` | 字符查找函数。从字符串 `abcdefg` 第 1 个字符开始查询，返回 ‘b’ 字符串第 1 次出现按的位置。 |

Reviewer: ```suggestion
| 字符查找函数 INSTR | `INSTR('abcdefg','b',1,1)` | `INSTR('abcdefg','b')` | 从字符串 'abcdefg' 第 1 个字符开始查询，返回 'b' 字符串第 1 次出现按的位置。 |
```

> | 字符查找函数 INSTR | `INSTR('stst', 's', 1, 2)` | `LENGTH(SUBSTRING_INDEX('stst','s',2))+1` | 字符查找函数。从字符串 'stst' 第一个字符开始查找，返回 's' 字符第 2 次出现的位置，查找非第一次出现的位置时使用。 |

Reviewer: ```suggestion
| 字符查找函数 INSTR | `INSTR('stst','s',1,2)` | `LENGTH(SUBSTRING_INDEX('stst','s',2))+1` | 从字符串 'stst' 第一个字符开始查找，返回 's' 字符第 2 次出现的位置，查找非第一次出现的位置时使用。 |
```

| 字符查找函数 | `INSTR('abcabc','b',2,1)` | `LOCATE('b','abcabc',2)` | 字符查找函数。从字符串 `abcabc` 第 2 个字符开始查询，返回 `b` 字符串第 1 次出现按的位置。 |
| 字符串引号识别区别 | 'a' | 'a' / "a" | Oracle 只能识别单引号，TiDB 能识别单引号与双引号。 |

> | null 与空串区别 | 不区分 null 和 '' | 区分 null 和 '' | Oracle 空串就是 null，TiDB 需要把空串转换为 null 数据。TiDB 中 null 和 '' 是有区别的。 |

Reviewer: ```suggestion
| NULL 与空串区别 | 不区分 NULL 和 '' | 区分 NULL 和 '' | Oracle 空串就是 NULL，TiDB 需要把空串转换为 NULL 数据。TiDB 中 NULL 和 '' 是有区别的。 |
```

> | 是否支持插入本表中查询出来的数据 | `INSERT INTO table1 VALUES (feild1,(select feild2 from table1 where...))` | `INSERT into table1 VALUES（feild1,(SELECT T.fields2 FROM table1 T WHERE...)` | TiDB 不支持在同一个表中先查这个表再更新该表。 |

Reviewer: ```suggestion
| 插入本表中查询出来的数据 | `INSERT INTO table1 VALUES (feild1,(SELECT feild2 FROM table1 WHERE...))` | `INSERT INTO table1 VALUES（feild1,(SELECT T.fields2 FROM table1 T WHERE...)` | TiDB 不支持在同一个表中先查这个表再更新该表。 |
```

| 获取日期之间间隔月份 | `MONTHS_BETWEEN(enddate,sysdate)` | `TIMESTAMPDIFF(MONTH,sysdate,enddate)` |  `MONTHS_BETWEEN` 函数返回两个日期之间的月份值，可以用 `TIMESTAMPDIFF` 替换，但是结果上会有误差，`TIMESTAMPDIFF` 只保留整数月，应该测试后按照业务看是否替换，另外注意参数位置相反。 |

> | 限制取前 n 条数据 | `ROWNUM <= 10` | `LIMIT 10` | 可以使用 LIMIT 等价代替，如：ROWNUM=1 使用 LIMIT 1 替换，hql 方式运行带 LIMIT 的 SQL 语句会出现错误，需要将 HIBERNATE 的运行方式改为 SQL 方式运行。 |

Reviewer: ```suggestion
| 限制取前 n 条数据 | `ROWNUM <= 10` | `LIMIT 10` | Oracle 中的 `ROWNUM` 可以使用 `LIMIT` 等价代替。Hibernate Query Language (HQL) 方式运行带 `LIMIT` 的 SQL 语句会出现错误，需要将 Hibernate 的运行方式改为 SQL 方式运行。 |
```

> | UPDATE 语句写法差异 | `UPDATE test1 SET(test1.name,test1.age)= (SELECT test2.name,test2.age FROM test2 WHERE test2.id=test1.id)` | `UPDATE test1,test2 SET test1.name=test2.name,test1.age=test2.age WHERE test1.id=test2.id` | TiDB 在多表更新的时候，需要在 SET 的时候把具体的字段更新关系都列出来。 |

Reviewer: ```suggestion
| UPDATE 语句 | `UPDATE test1 SET(test1.name,test1.age)= (SELECT test2.name,test2.age FROM test2 WHERE test2.id=test1.id)` | `UPDATE test1,test2 SET test1.name=test2.name,test1.age=test2.age WHERE test1.id=test2.id` | TiDB 在多表更新的时候，需要在 `SET` 的时候把具体的字段更新关系都列出来。 |
```

| 派生表别名 | `SELECT * FROM (SELECT * FROM test)` | `SELECT * FROM (SELECT * FROM test)` t | TiDB 多表查询的时候，每一个派生出来的表都必须有一个自己的别名。 |

> | MINUS 差集运算 | `SELECT * FROM t1 MINUS SELECT * FROM t2;` | `SELECT * FROM t1 EXCEPT SELECT * FROM t2;` | TiDB 不支持 MINUS，需要改写为 EXCEPT。 |

Reviewer: ```suggestion
| 差集运算 | `SELECT * FROM t1 MINUS SELECT * FROM t2;` | `SELECT * FROM t1 EXCEPT SELECT * FROM t2;` | TiDB 不支持 `MINUS`，需要改写为 `EXCEPT`。 |
```

> | 空值取别名差异 | `SELECT null AS xx FROM dual` | `SELECT '' AS xx FROM dual` | TiDB 数据库下，SQL 中字段直接为 NULL AS 的，在程序中运行会导致报错，需要改成 ''。NULL 与 '' 在 TiDB 中含义不同。 |

Reviewer: ```suggestion
| 空值取别名 | `SELECT null AS xx FROM dual` | `SELECT '' AS xx FROM dual` | NULL 与 '' 在 TiDB 中含义不同。Oracle 中 NULL 在 TiDB 中需要改写为 ''。|
```

> | 注释差异 |  --注释 |  -- 注释 | Oracle 的 -- 后面不需要空格，TiDB 的 -- 后面则需要有一个空格。 |

Reviewer: ```suggestion
| 注释 |  `--注释` |  `-- 注释` | Oracle 的 `--` 后面不需要空格，TiDB 的 `--` 后面需要有一个空格。|
```

> | 列转行函数差异 | `LISTAGG(concat(E.dimensionid,'---',E.DIMENSIONNAME),'***') within GROUP(ORDER BY  DIMENSIONNAME)` | `GROUP_CONCAT(concat(E.dimensionid,'---',E.DIMENSIONNAME) ORDER BY DIMENSIONNAME SEPARATOR '***')` | Oracle 中的 LISTAGG 需要改写为 TiDB 的 `GROUP_CONCAT` 函数；将一列字段合并为一行并根据 *** 符号进行分割。 |

Reviewer: ```suggestion
| 列转行函数 | `LISTAGG(concat(E.dimensionid,'---',E.DIMENSIONNAME),'***') within GROUP(ORDER BY  DIMENSIONNAME)` | `GROUP_CONCAT(concat(E.dimensionid,'---',E.DIMENSIONNAME) ORDER BY DIMENSIONNAME SEPARATOR '***')` | Oracle 中的 `LISTAGG` 需要改写为 TiDB 的 `GROUP_CONCAT` 函数；将一列字段合并为一行并根据 `***` 符号进行分割。 |
```

> | 分页查询 | `SELECT * FROM tables OFFSET 0 ROWS FETCH NEXT 2000 ROWS ONLY` | `SELECT * FROM tables LIMIT 2000 OFFSET 0` | 分页查询，OFFSET m 表示跳过 m 行数据，`FETCH NEXT n ROWS ONLY` 表示取 n 条数据，TiDB 使用 `LIMIT n OFFSET m` 进行等价改写。 |

Reviewer: ```suggestion
| 分页查询 | `SELECT * FROM tables OFFSET 0 ROWS FETCH NEXT 2000 ROWS ONLY` | `SELECT * FROM tables LIMIT 2000 OFFSET 0` | Oracle 中 `OFFSET m` 表示跳过 `m` 行数据，`FETCH NEXT n ROWS ONLY` 表示取 `n` 条数据。TiDB 使用 `LIMIT n OFFSET m` 等价改写 `OFFSET m ROWS FETCH NEXT n ROWS ONLY`。 |
```

> | 获取当前时间 `SYSTIMESTAMP` | `SYSTIMESTAMP` | `CURRENT_TIMESTAMP(6)` | 获取当前时间，时间值带微秒。 |

Reviewer: ```suggestion
| 获取当前时间（精确到微秒）| `SYSTIMESTAMP` | `CURRENT_TIMESTAMP(6)` | |
```

> | 特殊字符 ASCII 码值 | `CHR(n)` | `CHAR(n)` | ASCII 值转换函数，可将 ASCII 值转换为对应的字符, Oracle 中制表符 `CHR(9)`/换行符 `CHR(10)`/回车符 `CHR(13)` 对应 TiDB 中的 `CHAR(9)`/`CHAR(10)`/`CHAR(13)`。 |

Reviewer: ```suggestion
| ASCII 值转化为对应字符 | `CHR(n)` | `CHAR(n)` | Oracle 中制表符 `CHR(9)`/换行符 `CHR(10)`/回车符 `CHR(13)` 对应 TiDB 中的 `CHAR(9)`/`CHAR(10)`/`CHAR(13)`。 |
```

| 19 | 获取随机序列值 | SYS_GUID() | UUID() | 返回一个通用唯一识别码 (UUID)。 |
| 20 | nvl()函数 | nvl(key, val) | ifnull(key, val) | 如果该字段值为空，则返回 val 值，否则返回该字段的值。 |
| 21 | nal2()函数 nvl2() | nvl(key, val1, val2) nvl2() | if(key is null, val1, val2) | 如果该字段值非 NULL，则返回 val1 值，否则返回 val2 值。       |
| 22 | decode()函数 | DECODE(key, val1, val2, val3) | if(key = val1, val2, val3) | 如果该字段值对于 val1，则返回 val2，反之返回 val3。 |
| 23 | decode()函数 | DECODE(value, if1, val1, if2,val2,...,ifn, valn, val) | case when value=if1 then val1 when value=if2 then val2,,,when value=ifn then valn else val end | 当该字段值等于条件 1 时，返回 val1，等于条件 2 时，返回 val2… |
| 24 | 获取字符串长度 | length(str) | char_length(str) | 获取字符串长度。 |
| 25 | 截取字符串函数 | substr('abcdefg', 0, 2) => ab substr('abcdefg', 1, 2) => ab | substring('abcdefg', 0, 2) => 空 substring('abcdefg', 1, 2) => ab | 截取字符串，Oracle 中起始位置 0 与 1 作用一样，TiDB 中 0 开始截取为空，若需从头开始截全，则应从 1 开始，TiDB 支持 subString 和 subStr 函数，作用相同，不用修改。但是要注意下标，TiDB 必须从 1 开始。 |

> | 26 | 字符查找函数 instr | instr('abcdefg', 'b', 1, 1) | instr('abcdefg', 'b') | 字符查找函数。从字符串 ’abcabc’ 第 1 个字符开始查询，返回 ‘b‘ 字符串第 1 次出现按的位置。 |

Author: ```suggestion
| 26 | 字符查找函数 instr | instr('abcdefg', 'b', 1, 1) | instr('abcdefg', 'b') | 字符查找函数。从字符串 ’abcdefg’ 第 1 个字符开始查询，返回 ‘b‘ 字符串第 1 次出现按的位置。 |
```

---
title: Oracle 与 TiDB 函数对照表
summary: 了解 Oracle 与 TiDB 函数对照表。
---

# Oracle 与 TiDB 函数对照表

本文档提供了 Oracle 与 TiDB 的函数对照表，方便你根据 Oracle 函数查找对应的 TiDB 函数。

| 序号 | 语法差异点 | Oracle | TiDB  | 说明 |
|---|---|---|---|---|
| 1 | 数据类型转换函数 | to_number(key) | convert(key, dataType) | 转换表字段值数据类型，TiDB 支持 BINARY、CHAR、DATE、DATETIME、TIME、SIGNED INTEGER、UNSIGNED INTEGER、DECIMAL |
| 2 | 数据类型转换函数 | to_char(key) | convert(key, dataType) | 转换表字段值数据类型，TiDB 支持 BINARY、CHAR、DATE、DATETIME、TIME、SIGNED INTEGER、UNSIGNED INTEGER、DECIMAL |

> | 3 | 日期转换字符串函数 | to_char(sysdate,'yyyy-MM-dd hh24:mi:ss')       | date_format(now(), '%Y-%m-%d %H:%i:%s') | 日期类型转换字符型函数，TiDB 的年月日时分秒字符大小写必须严格按要求写。将原来的修改为 date_formate 修改为了修改为 date_format。 |

Author: 最后一句话要保留吗？

> | 4 | 日期转换字符串函数 | to_char(sysdate, 'yyyy-MM-dd')       | date_format(now(), '%Y-%m-%d') | 日期类型转换字符型函数。TiDB 的年月日时分秒字符大小写必须严格按要求走写。将原来的修改为 date_formate 修改为 date_format。 |

Author: ditto

| 6 | 字符串转换日期函数 | to_date('2021-05-28',   'yyyy-MM-dd hh24:mi:ss')       | str_to_date('2021-05-28',   '%Y-%m-%d%T') | 字符型转换日期型函数，TiDB 年月日时分秒字符大小写必须严格按要求写。 |
| 7 | Sysdate 关键字 | SYSDATE | NOW() | 获取系统当前时间。 |
| 8 | 获取两个日期相差的天数 | date1 - date2 | datediff(date1, date2) | 获取 date1 - date2 两个日期之间相差的天数，只能精确到天。 |
| 9 | 日期数据 +/- n 天 | dateVal + n | date_add(dateVal, INTERVAL n DAY) | 日期数据增加 `n` 天，`n` 可为负数 |
| 10 | 日期数据 +/- n 月 date_add() 函数替换 add_months() 函数 | add_months(dateVal, 1) | date_add(dateVal,   INTERVAL n MONTH) | 日期数据增加 `n` 月，`n` 可为负数。 |
| 11 | TRUNC()函数 | TRUNC(sysdate) | cast(now() as date) | 获取时间的（2019-07-26 00:00:00）格式返回值。Oracle 中的 TRUNC(sysdate) 只是截取到日，不会截取到时分秒，而 TiDB 中与之对应的截取日写法是 cast(now() as date)。 |
| 12 | TRUNC()函数 | TRUNC(sysdate) | date_format(now(),'%Y-%m-%d') | cast 数据库执行正常，程序中报错，可修改为 date_format。 |
| 13 | TRUNC()函数 | Trunc(sysdate,'mm') | date_add(curdate(),interval -day(curdate())+1 day)  | 获取当月第一天 |
| 14 | TRUNC 函数 | TRUNC(2.136) => 2 TRUNC(2.136, 2) => 2.14 | TRUNCATE(2.136, 0) => 2 TRUNCATE(2.136, 2) => 2.14 | 数据精度保留，直接截取相应小数位，不涉及四舍五入。 |
| 15 | 字符串连接 | a' \|\| 'b' | CONCAT('a', 'b') | 字符串拼接。 |

> | 16 | DELETE关键字 | DELETE   FROM tName t WWHER t.xxx = xxx; | DELETE FROM table_name WHERE xxx = xxx; | 删除语句，TiDB 不支持删除语句中对表起别名（有问题，tidb支持这种表别名写法）。 |

Author: 括号中的这句话要保留吗？

| 17 | 获取序列值 | sequenceName.nextVal | NEXTVAL(sequenceName) | 获取序列的下一个值。 |

> | 18 | 左/右关联 | SELECT \* FROM taba, tabb WHERE taba.id = tabb.id(+); SELECT \* FROM taba, tabb WHERE taba.id(+) = tabb.id; | SELECT \* FROM taba LEFT JOIN tabb ON taba.id = tabb.id; SELECT \* FROM taba RIGHT JOIN tabb ON taba.id = tabb.id;       | 关联查询时，TiDB 不支持使用 (+) 实现左/右关联，只能通过 left/right join 实现（原来例子中写反了左外连接和右外连接的顺序）。 |

Author: ditto

| 19 | 获取随机序列值 | SYS_GUID() | UUID() | 返回一个通用唯一识别码 (UUID)。 |
| 20 | nvl()函数 | nvl(key, val) | ifnull(key, val) | 如果该字段值为空，则返回 val 值，否则返回该字段的值。 |
| 21 | nal2()函数 nvl2() | nvl(key, val1, val2) nvl2() | if(key is null, val1, val2) | 如果该字段值非 NULL，则返回 val1 值，否则返回 val2 值。       |
| 22 | decode()函数 | DECODE(key, val1, val2, val3) | if(key = val1, val2, val3) | 如果该字段值对于 val1，则返回 val2，反之返回。 |
| 23 | decode()函数 | DECODE(value, if1, val1, if2,val2,...,ifn, valn, val) | case when value=if1 then val1 when value=if2 then val2,,,when value=ifn then valn else val end | 当该字段值等于条件 1 时，返回 val1，等于条件 2 时，返回 val2… |
| 24 | 获取字符串长度 | length(str) | char_length(str) | 获取字符串长度。 |
| 25 | 截取字符串函数 | substr('abcdefg', 0, 2) => ab substr('abcdefg', 1, 2) => ab | substring('abcdefg', 0, 2) => 空 substring('abcdefg', 1, 2) => ab | 截取字符串，Oracle 中起始位置 0 与 1 作用一样，TiDB 中 0 开始截取为空，若需从头开始截全，则应从 1 开始，TiDB 支持 subString 和 subStr 函数，作用相同，不用修改。但是要注意下标，TiDB 必须从 1 开始。 |

> | 26 | 字符查找函数 instr | instr('abcdefg', 'b', 1, 1) | instr('abcdefg', 'b') | 字符查找函数。从字符串 ’abcabc’ 第 1 个字符开始查询，返回 ‘b‘ 字符串第 1 次出现按的位置。 |

Author: ’abcabc’ 应该是 'abcdefg' 吧？

| 27 | 字符查找函数instr | instr('stst', 's', 1, 2) | LENGTH(SUBSTRING_INDEX('stst','s',2))+1 | 字符查找函数。从字符串 'stst' 第一个字符开始查找，返回 's' 字符第 2 次出现的位置，查找非第一次出现的位置时使用。 |
| 28 | 字符查找函数 | instr(‘abcabc’, 'b', 2, 1) | LOCATE(’b’，'abcabc’,2) | 字符查找函数。从字符串 ’abcabc’ 第 2 个字符开始查询，返回 ‘b‘ 字符串第 1 次出现按的位置。 |
| 29 | 字符串引号识别区别 | 'a' | 'a' / "a" | Oracle 只能识别单引号，TiDB 能识别单引号与双引号。 |

> | 30 | number数据类型精度问题 | 　 | 　 | 需要注意的数据类型：number 不手动写精度和标度，既可以支持小数也可以支持整型，需要单独判断。number 可以转成 decimal 或者 bigint。 |

Author: 中间空了两列，需要补充内容吗？

| 38 | select null as xx from dual | select   null as xx from dual | select '' as xx from dual | TiDB 数据库下，SQL 中字段直接为 null as 的，在程序中运行会导致报错，需要改成 ''。null 与 '' 在 TiDB 中含义不同。 |
| 39 | decimal 字段在 hibernate 中返回类型差异 | 0.00000000 在 Oracle 中返回 0 | 0.00000000 在 Oracle 中返回科学计数法 | TiDB 中小数类型如：0.00000000 返回到程序中后还是 0.00000000，导致在 BIGDECIMAL 进行科学计数法后会变成 0E-8，而 Oracle 则是直接返回为 0。 |
| 40 | 注释差异 |  --注释 |  -- 注释 | Oracle 的 -- 后面不需要空格，TiDB 的 -- 后面则需要有一个空格。 |
| 41 | 列转行函数差异 | listagg(concat(E.dimensionid,'---',E.DIMENSIONNAME),'***') within GROUP(ORDER BY  DIMENSIONNAME) | GROUP_CONCAT(concat(E.dimensionid,'---',E.DIMENSIONNAME) ORDER BY DIMENSIONNAME SEPARATOR '***') | Oracle 中的 listagg 需要改写为 TiDB 的 group_concat 函数；将一列字段合并为一行并根据 *** 符号进行分割。 |
| 42 | 分页查询 | select \* from tables OFFSET 0 ROWS FETCH NEXT 2000 ROWS ONLY | select \* from tables limit 2000 offset 0 | 分页查询，offset m 表示跳过 m 行数据，FETCH NEXT n ROWS ONLY 表示取 n 条数据，TiDB 使用 limit n offset m 进行等价改写。 |
| 43 | 获取当前时间 systimestamp | systimestamp | CURRENT_TIMESTAMP(6) | 获取当前时间，时间值带微秒。 |
| 44 | 特殊字符 ASCII 码值 | chr(n) | char(n) | ASCII 值转换函数，可将 ASCII 值转换为对应的字符, Oracle 中制表符 chr(9)/换行符 chr(10)/回车符 chr(13) 对应 TiDB 中的 char(9)/char(10)/char(13)。 |

> | 45 | Oracle 和 TiDB 排序 null 的顺序不同 | order by colum asc nulls first | order by colum asc | MySQL 和 TiDB 结：order by colum asc 时，null 默认被放在最前, order by colum desc 时，null 默认被放在最后。O：select \* from t1 order by name nulls first; 等价于 T：select \* from t1 order by NAME ; O：select \* from t1 order by name desc nulls last; 等价于 T：select \* from t1 order by name desc; O：select * from t1 order by name desc nulls first; 等价于 T：select \* from t1 order by  isnull(name) desc, name desc; O：select \* from t1 order by name asc nulls last; 等价于 T：select \* from t1 order by isnull(name), name; |

Author: MySQL 和 TiDB 总结是什么意思？
